doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,292 @@
+  The presence of the argument makes it obvious that errors _can_ happen, but you can easily omit the error check when you want to.
+- Adding error indications should not make the API less convenient to use.
+
+## Assumptions
+
+- In many cases, it is possible to avoid error cases through the preceding code path.
+  For example, a very common case is that functions return an error, if they receive a `NULL` pointer.
+  Often this case can be excluded by static analysis, i.e., because of the preceding code it is impossible that the pointer is `NULL`.
+
+  ```c
+  int foo (KeySet * ks, Key * key) {
+    if (key == NULL) {
+      return -1;
+    }
+
+    // [...] could lots of code, as long as the Key * key doesn't change
+
+    const Key * found1 = ksLookupByName (ks, "/foo", 0);
+    if (found1 != NULL) {
+      // no need to worry about key == NULL, because we checked earlier
+      // also no need to worry about ks == NULL, because the ksLookupByName would have returned NULL
+      const Key * found2 = ksLookup (ks, key, 0);
+    }
+  }
+  ```
+
+- It is not always necessary to report separate error codes, for _every_ error case.
+  In many cases, the caller can distinguish after the fact, e.g., `keySetName` fails when the given name is invalid and when the given `Key` has a read-only name.
+  This can be distinguished after the fact, by checking whether the `Key` has a read-only name with `keyIsLocked`.
+
+- Callers normally don't care about the exact error, unless they can do something about it.
+  This could be:
+
+  - reporting the error to a user
+  - switching to a different code path
+  - (during debugging) changing the code
+
+- Functions can be written in a such a way that they detect and report all errors without permanent effects.
+
+- The functions that cannot use a `Key * errorKey`, are side-effect-free.
+  Therefore, the only possible cause of errors are invalid arguments.
+
+## Considered Alternatives
+
+### Extra argument
+
+All functions that report errors take a `int * error` argument to which errors can be written.
+
+This makes the API much more convoluted to use.
+
+A benefit would be the return value remains fully usable.
+
+### Thread Local Storage
+
+Use a field in thread local storage (TLS) to store errors similar to `errno`.
+
+This leaves the API design entirely open.
+
+TLS was only standardized in C11, while Elektra currently only requires C99.
+
+This has all the same issues as `errno`: needs to be cleared to detect errors, can only store one error (per thread), etc.
+
+### Redesign API
+
+Redesign the API, so that every API function only has maximal one error case next to memory errors.
+
+- `nameNew(name) -> name_object` only has the error case of invalid name
+- `keyNew(name_object)` won't have error cases anymore
+- `keySetName(name_object)` only has the error case of read-only names
+
+This can be very limiting and in some cases makes it very hard to find out what the exact problem was.
```

Comment:
```suggestion
The advantage is that it is hard to be used wrongly.
Users can only call keySetName if they have a valid key name.
If keySetName fails nevertheless, we know that it must be because the name was read-only.
No further debugging would be needed.

This can be very limiting and in some cases makes it very hard to find out what the exact problem was.
```